src/backends/kimchi/mod.rs

```diff
@@ -688,28 +688,105 @@ impl Backend for KimchiVesta {
     }
 
     fn assert_eq_const(&mut self, cvar: &KimchiCellVar, cst: Self::Field, span: Span) {
-        self.add_generic_gate(
-            "constrain var - cst = 0 to check equality",
-            vec![Some(*cvar)],
-            vec![
-                Self::Field::one(),
-                Self::Field::zero(),
-                Self::Field::zero(),
-                Self::Field::zero(),
-                cst.neg(),
-            ],
+        // use permutation to check
+        let vars = vec![Some(*cvar)];
+        let coeffs = vec![
+            Self::Field::one(),
+            Self::Field::zero(),
+            Self::Field::zero(),
+            Self::Field::zero(),
+            cst.neg(),
+        ];
+        let note = "constrain var - cst = 0 to check equality";
+
+        // sanitize
+        assert!(coeffs.len() <= NUM_REGISTERS);
+        assert!(vars.len() <= NUM_REGISTERS);
+
+        // construct the execution trace with vars, for the witness generation
+        self.witness_table.push(vars.clone());
```

Comment:
you are using the witness table here, where we should not add anything to the witness table as we just want to wire things around. (Also, note to myself, variables are always constrained in circuit at this point? maybe we should differentiate between variable and allocated variable, cc @katat )

Reply:
I've removed the witness table additon.